File: python3/binary_tree_right_side_view.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        ans = []
        stack = [root]
        
        while stack:
            ans.append(stack[-1].val)
            temp = []
            for node in stack:
                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)
                    
            stack = temp
        
        return ans

File: golang/binary_tree_right_side_view.go
package golang

import (
	"golang/model"
)


func rightSideView(root *model.TreeNode) []int {
    var ans []int
    stack := []*model.TreeNode{root}
    
    if root == nil {
        return ans
    }
    
    for len(stack) > 0 {
        var temp []*model.TreeNode
        for i := range stack {
            if i == (len(stack)-1) {
                ans = append(ans, stack[i].Val)
            }
            if stack[i].Left != nil {
                temp = append(temp, stack[i].Left)
            }
            if stack[i].Right != nil {
                temp = append(temp, stack[i].Right)
            }
        }
        
        stack = temp
    }

    return ans
}